### PR TITLE
[STEP-174]: Ignoring utility variants, if not explicitly defined; Support multiple variants

### PR DESCRIPTION
**Pull request by lpusok**
https://bitrise.atlassian.net/browse/STEP-174

### What?
- Variants with suffixes `Classes`, `Resources`, `UnitTestClasses`, `AndroidTestClasses`, `AndroidTestResources` are ignored if the variants is not specifed.
- Multiple variants can be specified for the Step as an input parameter.

### Why?
- Based on [this](https://github.com/bitrise-steplib/bitrise-step-android-build/issues/12) issue, when the variant is not specified explicitly, the Step runs the Gradle tasks for all of the variants, which can result in invalid runs for the utility variants.
- Previously there was no option to set multiple variants, only one or all, which could cause poor execution time if there were several variants available.

### How?
- Filtering the utility variants will eliminate the invalid runs.
- Variants can be input in the following format: `value1\nvalue2` (you may find further info [here](https://devcenter.bitrise.io/contributors/create-your-own-step/#accepting-a-list-of-values-for-inputs)).

### CODE REVIEW

File: main_test.go [hunk 1]
```diff
@@ -91,4 +91,117 @@ func TestFilterVariants(t *testing.T) {
...
+		require.NoError(t, err)
+
+		expectedVariants := gradle.Variants{
+			"module1": []string{"variant1", "variant2"},
+		}
+
+		require.Equal(t, expectedVariants, filtered)
+	}
+
+	t.Log("exact match for module and multiple variants")
```
Comment (mateherber) [inline main_test.go:107]: I may miss something but this seems to be == with the one above

Comment (istvankovacs-bitrise) [reply main_test.go:107]: You are right! The curse of copy-paste. 😄

File: main.go [hunk 2]
```diff
@@ -92,6 +94,37 @@ func exportArtifacts(artifacts []gradle.Artifact, deployDir string) ([]string, e
 	return paths, nil
 }
 
+func filterUtilityVariants(variants []string) []string {
+	ignoredSuffixes := [...]string{"Classes", "Resources", "UnitTestClasses", "AndroidTestClasses", "AndroidTestResources"}
```
Comment (mateherber) [inline main.go:98]: What if somebody has a "classes" build type?

File: main.go [hunk 3]
```diff
@@ -102,23 +135,35 @@ func filterVariants(module, variant string, variantsMap gradle.Variants) (gradle
 		variantsMap = gradle.Variants{module: v}
 	}
 
-	// if variant not set: use all variants
+	// if variant not set: use all variants, except utility ones
 	if variant == "" {
+		for module, variants := range variantsMap {
+			variantsMap[module] = filterUtilityVariants(variants)
```
Comment (mateherber) [inline main.go:141]: Feels scary to modify a map while iterating over it. Even if we only change the values. Does go handle this nicely?

File: main.go [hunk 4]
```diff
@@ -92,6 +94,37 @@ func exportArtifacts(artifacts []gradle.Artifact, deployDir string) ([]string, e
 	return paths, nil
 }
 
+func filterUtilityVariants(variants []string) []string {
+	ignoredSuffixes := [...]string{"Classes", "Resources", "UnitTestClasses", "AndroidTestClasses", "AndroidTestResources"}
+	var filteredVariants []string
+
+	for _, v := range variants {
```
Comment (mateherber) [inline main.go:101]: This block is a perfect example why Go is not a productive language :D

in Kotlin it'd be `return variants.filter{variant -> ignoredSuffixes.none{variant.endsWith(it)}}`

File: main.go (hunk 2 above)
Comment (mateherber) [inline main.go:98]: Shouldn't we define this list as a const instead? Not in the func?

File: main.go (hunk 3 above)
Comment (lpusok) [reply main.go:141]: It is legal, adding/removing elements is more tricky tough. https://github.com/golang/go/issues/9926

File: main.go [hunk 5]
```diff
@@ -102,23 +135,35 @@ func filterVariants(module, variant string, variantsMap gradle.Variants) (gradle
...
 	filteredVariants := gradle.Variants{}
-	for m, variants := range variantsMap {
-		for _, v := range variants {
-			if strings.ToLower(v) == strings.ToLower(variant) {
-				filteredVariants[m] = append(filteredVariants[m], v)
+	for _, variant := range variants {
+		found := false
+		for m, moduleVariants := range variantsMap {
+			for _, v := range moduleVariants {
+				fmt.Println(v)
```
Comment (mateherber) [inline main.go:154]: this println is intentional?

File: step.yml [hunk 6]
```diff
@@ -42,9 +42,9 @@ inputs:
     opts:
       title: Variant
       summary: |
-        Set the variant that you want to build. To see your available variants please open your project in Android Studio and go in [Project Structure] -> variants section.
+        Set the variant(s) that you want to build. To see your available variants please open your project in Android Studio and go in [Project Structure] -> variants section.
       description: |
-        Set the variant that you want to build. To see your available variants please open your project in Android Studio and go in [Project Structure] -> variants section.
+        Set the variant(s) that you want to build. To see your available variants please open your project in Android Studio and go in [Project Structure] -> variants section.
```
Comment (lpusok) [inline step.yml:47]: Please add an example input in the description, and mention to use newline as seperator.

File: main.go [hunk 7]
```diff
@@ -92,6 +94,37 @@ func exportArtifacts(artifacts []gradle.Artifact, deployDir string) ([]string, e
 	return paths, nil
 }
 
+func filterUtilityVariants(variants []string) []string {
```
Comment (lpusok) [inline main.go:97]: It may be more accurate to call this filterNonUtilityVariants as that is returned :|

File: main.go [hunk 8]
```diff
@@ -92,6 +94,37 @@ func exportArtifacts(artifacts []gradle.Artifact, deployDir string) ([]string, e
...
+	}
+
+	return filteredVariants
+}
+
+func separateVariants(variantsAsOneLine string) []string {
+	variants := strings.Split(variantsAsOneLine, newLine)
+
+	for _, variant := range variants {
+		variant = strings.TrimSpace(variant)
```
Comment (lpusok) [inline main.go:122]: This deos not actually assign a new value in the slice :(. You need to use variants[i] = instead.

File: main.go [hunk 9]
```diff
@@ -102,23 +135,35 @@ func filterVariants(module, variant string, variantsMap gradle.Variants) (gradle
...
-	for m, variants := range variantsMap {
-		for _, v := range variants {
-			if strings.ToLower(v) == strings.ToLower(variant) {
-				filteredVariants[m] = append(filteredVariants[m], v)
+	for _, variant := range variants {
+		found := false
+		for m, moduleVariants := range variantsMap {
+			for _, v := range moduleVariants {
+				fmt.Println(v)
+				if strings.ToLower(v) == strings.ToLower(variant) {
```
Comment (lpusok) [inline main.go:155]: NOTE: You can use strings.EqualFold instead.

**Review by adborbas — DISMISSED** [review]

**Comment by adborbas** [comment]
Sorry did not see your comments.

File: main.go (hunk 2 above)
Comment (istvankovacs-bitrise) [reply main.go]: The algorithm will remove it too. We might modify the condition to ignore the exact matches and filter only if it is a "real" suffix? Wdyt?

Comment (mateherber) [reply main.go]: I'm not sure we can come up with a non-exploitable solution parsing just the names. This was just a thought-provoking question. I see we a lot of times rely on name / string parsing in the CI which can go south easily. I think "classes" build-type is not likely so we should handle it if we get a bug report :)

Comment (istvankovacs-bitrise) [reply main.go]: Sounds good to me, many thanks!

**Comment by lpusok** [comment]
LGTM, as author can not approve tough.

**Review by mateherber — APPROVED** [review]
LGTM

**Pull request merged**